lib/spack/llnl/util/filesystem.py

```diff
@@ -1156,7 +1156,9 @@ class HeaderList(FileList):
 
     # Make sure to only match complete words, otherwise path components such
     # as "xinclude" will cause false matches.
-    include_regex = re.compile(r'(.*)(\binclude\b)(.*)')
+    # Avoid matching paths such as <prefix>/include/something/detail/include,
+    # e.g. in the CUDA Toolkit which ships internal libc++ headers.
+    include_regex = re.compile(r'((?:(?!(?:\binclude\b)).)*)(\binclude\b)(.*)')
```

Comment:
```suggestion
    include_regex = re.compile(r'(.*?)(\binclude\b)(.*)')
```
Not a regex master (proven by the fact that I don't understand at a first glance what's going on here :laughing: ). By reading just the comment though I wonder if the simpler modifications above works fine. It basically says that the first match should be non-greedy.

Reply:
Wow, yes :D